README.md
# magicserver-clojure

A basic HTTP Server in CLojure.

Can handle GET and POST requests.

## Installation

Download the project and place your main app file in /src/magicserver_clojure.

## Usage

To start the server at port 8888

    $ lein run 8888
## How to use:

Static files have to be enclosed in 'views' directory under root.

```
/
views/
    js/
    img/
    css/
```

To map the dynamic pages, use the function `server.add-route()` which takes 3 parameters

1. HTTP Method.
2. Requested path.
3. Function that would return the dynamic content.

Eg:

```
(defn home [request response]
  (server.send-html-handler request response content))
  
(server.add-route 'get' '/' home)
```

To send html or json data response, use the following functions `server.send-html-handler`  and `server.send-json-handler` which take 3 arguments

1. request
2. response
3. requested HTML/JSON content

### Bugs

Raise a issue or send a pull request if you find any bugs.

## License

Copyright © 2017 p4v4n

Distributed under the Eclipse Public License (version 1.0)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-03-31
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-03-31
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/magicserver-clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/magicserver-clojure/compare/0.1.0...0.1.1


test/magicserver_clojure/core_test.clj
(ns magicserver-clojure.core-test
  (:require [clojure.test :refer :all]
            [magicserver-clojure.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/magicserver_clojure/main.clj
(ns magicserver-clojure.main)
(use 'magicserver-clojure.sessions)
(use 'magicserver-clojure.server)


(defn add-route [method path func]
    (swap! ROUTES assoc method (assoc (@ROUTES method) path func)))


(defn home [request response]
    (let [session-cookie (get-session request)
    	userid (if session-cookie (session-cookie "userid"))]
    (if userid
    	(send-html-handler request response (format (slurp "./views/dashboard.html") userid))
        (send-html-handler request response (slurp "./views/index.html")))))


(defn login [request response]
    (let [username ((request "content") "username")
          password ((request "content") "password")]
         (add-session request {"userid" username})
        (send-html-handler request response (format (slurp "./views/dashboard.html") username))))

(defn logout [request response]
	(del-session request)
    (send-html-handler request response (slurp "./views/index.html")))

(add-route "get" "/" home)
(add-route "post" "/login" login)
(add-route "post" "/logout" logout)


src/magicserver_clojure/sessions.clj
(ns magicserver-clojure.sessions)


(def SESSIONS (atom {}))


(defn uuid [] (str (java.util.UUID/randomUUID)))


(defn session-handler [request response]
    (let [browser-cookies (request "Cookie")
          sid (if browser-cookies (browser-cookies "sid"))
          session-check (if sid (@SESSIONS sid))]
        (if (and browser-cookies sid session-check)
            response
            (let [cookie (uuid)]
                (do (swap! SESSIONS assoc cookie {})
                (assoc response "Set-Cookie" (str "sid=" cookie)))))))


(defn add-session [request content]
    (let [browser-cookies (request "Cookie")]
        (if (browser-cookies "sid")
            (swap! SESSIONS assoc (browser-cookies "sid") content))))


(defn get-session [request]
    (let [browser-cookies (request "Cookie")]
        (if (and browser-cookies (browser-cookies "sid"))
            (@SESSIONS (browser-cookies "sid")))))


(defn del-session [request]
    (let [browser-cookies (request "Cookie")]
        (if (browser-cookies "sid")
            (swap! SESSIONS dissoc (browser-cookies "sid")))))


src/magicserver_clojure/server.clj
;; Magic Server in Clojure


(ns magicserver-clojure.server)
(use 'magicserver-clojure.sessions)
(use '[clojure.string])
(use '[clojure.data.json :as json])
(require '[clojure.java.io :as io])
(import '[java.net ServerSocket])


(def CONTENT-TYPE {"html" "text/html", "css" "text/css", "js" "application/javascript",
                   "jpeg" "image/jpeg", "jpg" "image/jpg", "png" "image/png", "gif" "image/gif",
                   "ico" "image/x-icon", "text" "text/plain", "json" "application/json"})


(def test-get-str "GET / HTTP/1.1\r\nHost: localhost:8888\r\nConnection: keep-alive\r\nCache-Control: max-age=0\r\nUpgrade-Insecure-Requests: 1\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/56.0.2924.76 Chrome/56.0.2924.76 Safari/537.36\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r\nDNT: 1\r\nAccept-Encoding: gzip, deflate, sdch, br\r\nAccept-Language: en-GB,en-US;q=0.8,en;q=0.6\r\n\r\n")
(def test-post-str "POST /submit HTTP/1.1\r\nHost: localhost:8888\r\nConnection: keep-alive\r\nContent-Length: 31\r\nCache-Control: max-age=0\r\nOrigin: http://localhost:8888\r\nUpgrade-Insecure-Requests: 1\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/56.0.2924.76 Chrome/56.0.2924.76 Safari/537.36\r\nContent-Type: application/x-www-form-urlencoded\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r\nDNT: 1\r\nReferer: http://localhost:8888/\r\nAccept-Encoding: gzip, deflate, br\r\nAccept-Language: en-GB,en-US;q=0.8,en;q=0.6\r\n\r\n")
(def test-post-str2 "POST /submit HTTP/1.1\r\nHost: localhost:8888\r\nConnection: keep-alive\r\nContent-Length: 31\r\nCache-Control: max-age=0\r\nOrigin: http://localhost:8888\r\nUpgrade-Insecure-Requests: 1\r\nUser-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/57.0.2987.98 Chrome/57.0.2987.98 Safari/537.36\r\nContent-Type: application/x-www-form-urlencoded\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r\nDNT: 1\r\nReferer: http://localhost:8888/\r\nAccept-Encoding: gzip, deflate, br\r\nAccept-Language: en-GB,en-US;q=0.8,en;q=0.6\r\n\r\nfirstname=Pavan&lastname=Mantha\r\n\r\n")


;;Parsers

(defn get-http-header [data]
    (if (re-find #"\r\n\r\n" data)
        (split data #"\r\n\r\n" 2)))


(defn process-request-first-line
    [input]
    (zipmap ["method" "path" "protocol"] (split input #"\s")))


(defn header-parser [input]
    (let [header-list (split input #"\r\n")]
        (loop 
            [header-map (process-request-first-line (first header-list))
            re (rest header-list)]
            (if (empty? re)
                (if (header-map "Cookie") 
                    (assoc header-map "Cookie" {"sid" (subs (header-map "Cookie") 4)}) 
                    header-map)
                (recur
                    (assoc 
                        header-map
                        (first (split (first re) #":\s+"))
                        (second (split (first re) #":\s+")))
                    (rest re))))))


(def date (.toString (java.util.Date.)))


(defn response-stringify [response]
    (let [keys-needed (filter #(and (not= % "status") (not= % "Content")) (keys response))
          response-string (str (response "status") "\r\n")]
         (loop [re response-string
                h-keys keys-needed]
                (if (empty? h-keys)
                    (str re "\r\n" (response "Content") "\r\n\r\n")
                    (recur (str re (first h-keys) ": " (response (first h-keys)) "\r\n") (rest h-keys))))))

;;Handlers

(defn response-handler [request response]
    (response-stringify (assoc response "Date" date "Connection" "close" "Server" "magic-server-clojure")))


(defn ok-200-handler [request response]
    (if (and (response "Content") (response "Content-type"))
        (response-handler request (assoc response "status" "HTTP/1.1 200 OK" "Content-Length" (str (count (response "Content")))))
        (response-handler request (assoc response "status" "HTTP/1.1 200 OK"))))


(defn err-404-handler [request response]
    (response-handler request (assoc response "status" "HTTP/1.1 404 Not Found" "content" "Content Not Found" "Content-type" "text/html")))


(defn send-html-handler [request response content]
    (if (not (empty? content))
        (ok-200-handler request (assoc response "Content" content "Content-type" "text/html"))
        (err-404-handler request response)))


(defn send-json-handler [request response content]
    (if (not (empty? content))
        (ok-200-handler request (assoc response "content" (json/write-str content) "Content-type" "application/json"))
        (err-404-handler request response)))


(defn static-file-handler [request response]
    (try
        (let [file-data (slurp (str "/public" (request "path")))
              content-type (last (split (request "path") #"."))]
            (ok-200-handler request (assoc response "content" file-data "Content-type" (CONTENT-TYPE content-type))))
        (catch Exception e (err-404-handler request response))))


(defn form-parser [request]
    (let [content-type (request "Content-Type")
          boundary (last (split content-type #"; "))
          boundary-value (str "--" (second (split boundary #"=")))
          content-list (split (request "body") (re-pattern boundary-value))]
        (loop [c-l content-list form {}]
            (if (empty? c-l)
                (assoc request "boundary" boundary-value "form" form)
                (let [form-data (split (first content-list) #"\r\n\r\n" 2)
                      form-header (split (first form-data) #"\r\n")
                      form-body (second form-data)
                      form-header-dict (into {} (map #(let [[k v] (clojure.string/split % #": " 2)] {k v}) form-header))
                      content-items (split (form-header-dict "Content-Disposition") #"; ")
                      data (into {} (map #(let [[k v] (clojure.string/split % #"=" 2)] {k v}) content-items))
                      new-data (assoc data "body" form-body)]
                    (recur (rest c-l) (assoc form (data "name") new-data)))))))


(defn parse-fields [body]
    (let [body-list (split body #"&")]
        (loop [body-dict {} b-list body-list]
            (if (empty? b-list)
                body-dict
                (let [[k v] (split (first b-list) #"=" 2)](recur (assoc body-dict k v) (rest b-list)))))))


(def ROUTES (atom {"get" {} "post" {} "put" {} "delete" {}}))


(defn get-handler [request response]
    (try
        (((@ROUTES "get") (request "path")) request response)
        (catch Exception e (static-file-handler request response))))


(defn post-handler [request response]
    (try
        (if (re-find #"multipart" (request "Content-Type"))
            (((@ROUTES "post") (request "path")) (assoc (form-parser request) "content" ((form-parser request) "form")) response)
            (((@ROUTES "post") (request "path")) (assoc request "content" (parse-fields (request "body"))) response))
        (catch Exception e (err-404-handler request response))))


(def METHOD {"GET" get-handler "POST" post-handler})


(defn method-handler [request response]
    ((METHOD (request "method")) request response))


(defn head-handler [request response]
    (response-handler request (assoc response "content" "")))


(defn request-handler [request]
    (method-handler request (session-handler request {})))


(defn worker [data]
    (let [[header-str body-str] (get-http-header data)]
        (request-handler (assoc (header-parser header-str) "body" (trim body-str)))))


(defn check-for-content-length [input]
    (let [header-dict (header-parser (trim input))
          content-length (header-dict "Content-Length")]
        (if content-length
           (Integer/parseInt content-length))))


(defn receive [socket]
    (let [x (io/reader socket)
          request-header (loop [new-line (.readLine x) 
                                final-string ""]
                                (if (empty? new-line)
                                    (str final-string "\r\n")
                                    (recur (.readLine x) (str final-string new-line "\r\n"))))
          content-length (check-for-content-length request-header)
          complete-request (if content-length 
                               (let [body (char-array content-length)]
                                   (.read x body 0 content-length)
                                   (str request-header (apply str body)))
                                request-header)]
    complete-request))


(defn send [socket msg]
    (let [writer (io/writer socket)]
        (.write writer msg)
        (.flush writer)))


(defn serve-persistent [port handler]
  (let [running (atom true)]
    (future
      (with-open [server-sock (ServerSocket. port)]
        (while @running
          (with-open [sock (.accept server-sock)]
             (let [msg-in (receive sock)
                  a (prn "Request: " msg-in)
                  msg-out (handler msg-in)
                  b (prn "Response: " msg-out)]
             (send sock msg-out))))))
    running))
